invoices/invoiceregistry.go

```diff
@@ -775,6 +781,67 @@ func (i *InvoiceRegistry) processKeySend(ctx invoiceUpdateCtx) error {
 	return nil
 }
 
+// processAMP just-in-time inserts an invoice if this htlc is a keysend
+// htlc.
+func (i *InvoiceRegistry) processAMP(ctx invoiceUpdateCtx) error {
+	// AMP payments MUST also include an MPP record.
+	if ctx.mpp == nil {
+		return errors.New("no MPP record for AMP")
+	}
+
+	// Create an invoice for the total amount expected, provided in the MPP
+	// record.
+	amt := ctx.mpp.TotalMsat()
+
+	// Set the TLV and MPP optional features on the invoice. We'll also make
+	// the AMP features required so that it can't be paid by legacy or MPP
+	// htlcs.
+	rawFeatures := lnwire.NewRawFeatureVector(
+		lnwire.TLVOnionPayloadOptional,
+		lnwire.PaymentAddrOptional,
+		lnwire.AMPRequired,
+	)
+	features := lnwire.NewFeatureVector(rawFeatures, lnwire.Features)
+
+	// Use the minimum block delta that we require for settling htlcs.
+	finalCltvDelta := i.cfg.FinalCltvRejectDelta
+
+	// Pre-check expiry here to prevent inserting an invoice that will not
+	// be settled.
+	if ctx.expiry < uint32(ctx.currentHeight+finalCltvDelta) {
+		return errors.New("final expiry too soon")
+	}
+
+	// We'll use the sender-generated payment address provided in the HTLC
+	// to create our AMP invoice.
+	payAddr := ctx.mpp.PaymentAddr()
+
+	// Create placeholder invoice.
```

Comment:
After looking at lnd flame graphs for some time, I now realize that first creating and then updating the invoice is not ideal for maximum throughput. Everything database shows up prominently on the graph. Yes, for multiple shards it matters less, because multiple updates are needed anyway (although - isn't it possible to keep the invoice states fully in-memory until the set is complete? the htlcs are replayed anyway after a restart). But I can imagine that the majority of the cases will be successful single shard AMP payments. Especially when AMP turns out to be the platform of choice for implementing money streaming.

Reply:
> although - isn't it possible to keep the invoice states fully in-memory until the set is complete? the htlcs are replayed anyway after a restart

I guess we do it somewhat for record keeping purposes as well, since by updating it incrementally we can track the arrival, settle, and thus the "hodl" time for a given invoice across all the HTLCs. At a glance though, if we're willing to sacrifice that information, or collect it in an alternative manner (that may potentially be more lossy), I think the same potential optimization can be applied to MPP given that we'll also update the invoice incrementally (the state) as new HTLCs start to arrive.

Reply:
This was also considered here: https://github.com/lightningnetwork/lnd/pull/5108/files/7a87a4c6627b6727facdfe1eb52aaeb5353f650a#r604347123 

Worthy of a follow up if we don't think it's useful to know the history of prior failed attempts of a spontaneous payment.